---
title: "Analyzing NanoString nCounter Data with the NanoTube"
author: "Caleb A Class"
runtime: shiny
output: html_document:
    toc: true
    toc_depth: 2
---

```{r prep, include = FALSE}

library(knitr)

```

# Abstract


# A basic workflow
`processNanoStringData` allows you to read in nCounter expression data (from
RCC files or in tabular form) and conduct basic normalization and quality
control checks in one step. We use example data from GEO data series 
GSE117751 (Lundy et al, 2018).

```{r Basic1}

library(NanoTube)

example_data <- system.file("extdata", "GSE117751_RAW", package = "NanoTube")



```

You can then run differential expression analysis using Limma. Functions are
also provided to allow for analysis using NanoStringDiff.

```{r Basic2}

```

DE results can be viewed or exported using `makeDiffExprFile`.

```{r Basic3}

```

DE results can also be used as input for gene set analysis in the `fgsea` package, 
using `limmaToFGSEA` or `nsdiffToFGSEA`.

```{r Basic4}

```

Gene set analysis results can be exported or plotted.

```{r Basic5}

```



# Processing Data
## From RCC files
One Reporter Code Count (RCC) file is generated by the nCounter instrument
for each sample, containing the counts for each gene and control in the 
codeset. It also includes some basic quality control (QC) metrics that can
by imported by NanoTube. Two functions are provided in this package: 
`read_rcc`, which reads in a single RCC file, and `read_merge_rcc` which reads
in a vector of RCC file names and combines the data. `processNanostringData`
also has the 

## From Other files



# Quality Control
NanoTube provides the quality control metrics recommended for NanoString
nCounter data. The raw NanoString data can be loaded for QC using the 
`normalization = "none"` option of `processNanostringData`.

```{r qc1}

dat <- processNanostringData(example_data, normalization = "none",
                             includeQC = TRUE)

```
Basic QC and cartridge data are loaded in from the RCC files if `includeQC` is set to TRUE.

```{r qc2}

head(pData(dat)[,-1])

```

Positive QC statistics can be calculated and presented as a table. This includes
the positive scaling factors and R-squared values for the expected vs. observed
positive control counts. NanoString recommends positive scaling factors between
0.3 and 3, and R-squared values greater than 0.95. Samples with values outside
these recommendations should be investigated further. 

```{r qc3}

posQC <- positiveQC(dat)

kable(head(posQC$tab), 
      row.names = FALSE, format = "html")

```

Positive control genes can be plotted for all samples (default), or a specified
subset of samples (specified by column index, or sample names).

```{r qc4}

posQC2 <- positiveQC(dat, samples = 1:6)

posQC2$plt

```


